compiler/rustc_trait_selection/src/solve/trait_goals/structural_traits.rs

```diff
@@ -182,18 +182,29 @@ pub(super) fn instantiate_constituent_tys_for_copy_clone_trait<'tcx>(
     }
 }
 
+#[derive(Debug)]
+pub(crate) enum CallableMode {
+    /// Check that the given type can implement a `Fn*` trait.
+    Fn(ty::ClosureKind),
+    /// Do not perform any signature checks, just extract the the tupled arguments and the output.
+    Callable,
+}
+
 // Returns a binder of the tupled inputs types and output type from a builtin callable type.
+#[instrument(level = "trace", skip(tcx), ret)]
 pub(crate) fn extract_tupled_inputs_and_output_from_callable<'tcx>(
     tcx: TyCtxt<'tcx>,
     self_ty: Ty<'tcx>,
-    goal_kind: ty::ClosureKind,
+    mode: CallableMode,
 ) -> Result<Option<ty::Binder<'tcx, (Ty<'tcx>, Ty<'tcx>)>>, NoSolution> {
+    use CallableMode::*;
     match *self_ty.kind() {
         // keep this in sync with assemble_fn_pointer_candidates until the old solver is removed.
         ty::FnDef(def_id, substs) => {
             let sig = tcx.fn_sig(def_id);
-            if sig.skip_binder().is_fn_trait_compatible()
-                && tcx.codegen_fn_attrs(def_id).target_features.is_empty()
+            if matches!(mode, Callable)
```

Comment:
I think it's easier to understand if we just match on `(*self_ty.kind(), mode)`

Reply:
I did that for the closures, but I don't see how we can improve anything for `FnDef` or `FnPtr`.